lib/msf/core/rhosts_walker.rb

```diff
@@ -0,0 +1,158 @@
+        set_backtrace(cause.backtrace) if cause
+      end
+    end
+
+    def initialize(value = '', datastore = Msf::ModuleDataStore.new(nil))
+      @value = value
+      @datastore = datastore
+    end
+
+    #
+    # Iterate over the valid rhosts datastores. This can be combined Calling `#valid?` beforehand to ensure
+    # that there are no invalid configuration values, as they will be ignored by this method.
+    #
+    # @yield [Msf::DataStore] Yields only _valid_ rhost values.
+    def each(&block)
+      return unless @value
+      return unless block_given?
+
+      parse(@value, @datastore).each do |result|
+        block.call(result) if result.is_a?(Msf::DataStore)
+      end
+
+      nil
+    end
+
+    # Count the _valid_ datastore permutations for the current rhosts value. This count will
+    # ignore any invalid values.
+    #
+    # @return [Integer]
+    def count
+      to_enum.count
+    end
+
+    #
+    # Retrieve the list of errors associated with this rhosts walker
+    # @yield [Msf::RhostsWalker::Error] Yields only invalid rhost values.
+    def errors(&block)
+      return unless @value
+      return unless block_given?
+
+      parse(@value, @datastore).each do |result|
+        block.call(result) if result.is_a?(Msf::RhostsWalker::Error)
+      end
+
+      nil
+    end
+
+    #
+    # Indicates that the rhosts value is valid and iterable
+    #
+    # @return [Boolean] True if all items are valid, and there are at least some items present to iterate over. False otherwise.
+    def valid?
+      parsed_values = parse(@value, @datastore)
+      parsed_values.all? { |result| result.is_a?(Msf::DataStore) } && parsed_values.count > 0
+    rescue StandardError => e
+      elog('rhosts walker invalid', error: e)
+      false
+    end
+
+    #
+    # Parses the input rhosts string, and yields the possible combinations of datastore values.
+    #
+    # @param value [String] the http string
+    # @return [Enumerable<Msf::DataStore|StandardError>] The calculated datastore values that can be iterated over for
+    #   enumerating the given rhosts, or the error that occurred when iterating over the input
+    def parse(input, datastore)
+      Enumerator.new do |results|
+        values = input.to_s.split(', ').map { |line| line.split(' ') }.flatten
+        values.each do |value|
+          if (value =~ %r{^file://(.*)}) || (value =~ /^file:(.*)/)
+            file = Regexp.last_match(1)
+            File.read(file).each_line(chomp: true) do |line|
+              parse(line, datastore).each do |result|
+                results << result
+              end
+            end
+          elsif value.start_with?('http:') || value.start_with?('https:')
+            http_options = parse_http_uri(value, datastore)
+            Rex::Socket::RangeWalker.new(http_options['RHOSTS']).each_ip do |ip|
+              results << datastore.merge(http_options.merge('RHOSTS' => ip))
+            end
+          elsif value =~ /^cidr:(.*)/
```

Comment:
We might want this regex to be a little more specific like: `/^cidr:(\/\d+)/` to ensure that it's at least as slash-number. Also FYI, if the bitmask is out of range, I'm pretty sure that `Rex::Socket::RangeWalker` will silently [drop it](https://github.com/rapid7/rex-socket/blob/master/lib/rex/socket/range_walker.rb#L95). Since you're doing better error handling now, it may be beneficial to also check that the bitmask is within the permissible range based on the address's family type.

Reply:
The RW drop thing is a bug IMO, though we probably have things depending on that behavior would would be nice to clean up.

Reply:
I've added validation below this line. I'd probably want to expose the existing validation from rex-socket long-term, so I've put in an initial regex check for now